### PR TITLE
Implement data storage and add it to Player class

### PR DESCRIPTION
**Pull request by Rochet2**
**Problem:**
Modules need to store data for a specific player.
They do so by making a map that maps player guid to some data.
This is not safe when hooks such as PvP kill can be called from two threads at same time and would require locks for every access to the map.

Example module that has the issue: https://github.com/azerothcore/mod-transmog/blob/master/src/Transmogrification.h#L42-L43
Btw I dont think null mutex actually acts as a lock in ace. Could be wrong though.

**Solutions:**
Each module would need to use locks, but they would reduce concurrency by doing so.
Better concurrency is achieved by storing the data directly to the Player class. However that means module makers need to edit Player.h.

**This solution:**
Test online: https://repl.it/repls/ShadyMajesticOcelot
Code: https://pastebin.com/mtimZJHG

The solution is to implement a storage to player class that can store data without knowing the type of it, but can safely return a pointer to it with the correctly requested type. This allows modules to store any data to player without editing player files.

Basically we make a container of pointers and use dynamic cast to convert them to correct user defined type.
Failing the casting or not finding an element results in nullptr.
Memory is managed automatically for the stored data so you can just create, store and forget.
Compile time checks are in place to ensure that set/get objects are derived from required base class. (without this dynamic cast could result in undefined behavior or we could see crash/ub during runtime)

Example of usage:
```c++
#include "DataMap.h"

class MyClass : public DataMap::Base
{
    // anything here
};

// store
player.DataMap.Set("modulename", new MyClass());
// get
if (MyClass* myclass = player.DataMap.Get<MyClass>("modulename"))
    myclass->stuff(); // Do stuff with myclass
// deleted when player logs out
```

The same system could be used per WorldObject for example to allow creature, gameobject, player specific storages and for maps too.

**Target branch(es):** 1.x/2.x etc.

**Issues addressed:** Closes #

**Tests performed:** (Does it build, tested in-game, etc)
Compiles

**Known issues and TODO list:**

- [ ] Test / use it with something
- [ ] Evaluate if it is wanted
- [ ] How to get ppl to use it / know about it?
- [ ] Use pimpl http://en.cppreference.com/w/cpp/language/pimpl

### CODE REVIEW

File: src/common/Utilities/DataMap.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license: http://github.com/azerothcore/azerothcore-wotlk/LICENSE-GPL2
```
Comment (ShinDarth) [inline src/common/Utilities/DataMap.h]: new files don't require to be released under GPL v2. You can use AGPL here

**Comment by Yehonal** [comment]
It's amazing Idea @Rochet2 ! i was trying to create custom uint32 vectors for world objects to store custom information to be reused by modules. But it's more advanced ofc.  

i'll test with some use-cases. And we should create a documentation about it then.

**Comment by Yehonal** [comment]
@Rochet2 i would accept this PR and starting to test it on custom XP rate module, but before....could you move `DataMap DataMap;` , from Player.h to Object.h please?

**Comment by Rochet2** [comment]
A related commit: https://github.com/azerothcore/azerothcore-wotlk/commit/d85e55dd9b521a24f901d93c34f736ff69b89af5

**Comment by Rochet2** [comment]
A lot of people asked how to use this. Here is an example.

However it also shows that some convenience functions should be implemented.
For example Erase instead of Set with a nullptr and GetDefault to get an object that is default constructed if one doesnt exist. (similar behavior to std::map, but we want type safety)

https://gist.github.com/Rochet2/8fd91f301a255865cb35a5d12341f7a2

**Comment by Rochet2** [comment]
I would like to note that I changed the API slightly, documented the code and implemented the convenience functions in https://github.com/azerothcore/azerothcore-wotlk/commit/0a3be94815675f8a12a5e04824d749c7db8041f0

**Pull request merged**